docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: example
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
+    "foo.example.com":
+    - url: "/foo/*"
+      service:
+        name: foosvc
+        namespace: default
+        port: 80
+    - url: "/bar/*"
+      service:
+        name: barsvc
+        namespace: default
+        port: 80
+    "bar.example.com"
+    - url: "/foobar/*"
+        service:
+          name: foobarsvc
+          namespace: default
+          port: 80
+status:
+  # include host here when we have DNS figured out.
+  address: 104.101.11.39
+```
+
+Put more rigidly:
+
+```go
+// IngressPoint represents a point for ingress traffic.
+type IngressPoint struct {
+	TypeMeta
+	ObjectMeta
+	Spec IngressPointSpec
+	Status IngressPointStatus
+}
+
+// IngressPointSpec describes the ingressPoint the user wishes to exist.
+type IngressPointSpec struct {
+	Host    string
+	IngressPoint map[string][]Path
+}
+
+// IngressPointStatus describes the current state of an ingressPoint.
+type IngressPointStatus struct {
+	Address string
+}
+
+// ServiceRef is a reference to a single service:port.
+type ServiceRef struct {
+	Name      string
+	Namespace string
+	Port      int64
```

Comment:
I think we should be explicit here - load balancers may want to only read
ingress and endpoints (and skip the service).  Also load balancers may have
limitations on types of ports (doesn't support http with connection
upgrade) and so we should try to encourage an explicit desire from users.

The numbers of ports is potentially multiple, but for now requiring
separate ingress rules is not burdensome (especially when load balancers
can distinguish between which ingress they wish to serve).

On Sep 13, 2015, at 12:15 AM, Alex Robinson <notifications@github.com>
wrote:

In docs/proposals/loadbalancing.md
<https://github.com/kubernetes/kubernetes/pull/12827#discussion_r39344298>:

> +// IngressPointSpec describes the ingressPoint the user wishes to exist.
> +type IngressPointSpec struct {
> +	Host    string
> +	IngressPoint map[string][]Path
> +}
> +
> +// IngressPointStatus describes the current state of an ingressPoint.
> +type IngressPointStatus struct {
> +	Address string
> +}
> +
> +// ServiceRef is a reference to a single service:port.
> +type ServiceRef struct {
> +	Name      string
> +	Namespace string
> +	Port      int64

Right, shouldn't this be an IntOrString to support either numeric or named
ports?

Also, is Port required if the service only has one port?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12827/files#r39344298>.